lib/api/formtools/grid.js

```diff
@@ -0,0 +1,263 @@
+
+                    if (filterErr) {
+                        return filtersDone(filterErr);
+                    }
+
+                    settings.filters[fieldName].formControls = result;
+
+                    return filtersDone(null);
+
+                });
+
+            }, function (filterErr) {
+
+                return cb(filterErr, settings.filters);
+
+            });
+
+        });
+
+    }
+
+    function getSearchFilters(fieldNames, data, modelName, cb) {
+
+        if (!fieldNames || !Array.isArray(fieldNames) || !fieldNames.length) {
+            throw new Error('fieldNames is required and must be of type array and cannot be emptied.');
+        }
+
+        if (!modelName || typeof(modelName) !== 'string' || !modelName.length) {
+            throw new Error('modelName is required and must be of type string and cannot be emptied.');
+        }
+
+        if (!data || typeof(data) !== 'object') {
+            throw new Error('data is required and must be of type object.');
+        }
+
+        var filters = {},
+            model = linz.get('models')[modelName];
+
+        async.waterfall([
+
+            function (callback) {
+                return getSettings(modelName, callback);
+            }
+
+        ], function (err, settings) {
+
+            if (err) {
+                return cb(err);
+            }
+
+            if (!Object.keys(settings.filters).length) {
+                return cb(null);
+            }
+
+            async.each(fieldNames, function (fieldName, filtersDone) {
+
+                if (!data[fieldName]) {
+                    return filtersDone(null);
+                }
+
+                // call the filter renderer and update the content with the result
+                settings.filters[fieldName].filter.filter(fieldName, data, function (filterErr, result) {
+
+                    if (filterErr) {
+                        return filtersDone(filterErr);
+                    }
+
+                    filters = model.addSearchFilter(filters, result);
+
+                    return filtersDone(null);
+
+                });
+
+            }, function (filterErr) {
+
+                if (filterErr) {
+                    return cb(filterErr);
+                }
+
+                // consolidate filters into query
+                filters = model.setFiltersAsQuery(filters);
+
+                return cb(filterErr, filters);
+            });
+
+        });
+
+    }
+
+    function getActiveFilters (selectedFilters, data, modelName, cb) {
+
+        if (!selectedFilters || !Array.isArray(selectedFilters) || !selectedFilters.length) {
+            throw new Error('selectedFilters is required and must be of type array and cannot be emptied.');
+        }
+
+        if (!modelName || typeof(modelName) !== 'string' || !modelName.length) {
+            throw new Error('modelName is required and must be of type string and cannot be emptied.');
+        }
+
+        if (!data || typeof(data) !== 'object') {
```

Comment:
I'm assuming that you want an object here, and not an Array. If so, this could be written as `if (!(data instanceof Object && !Array.isArray(data))) {`. Using instanceof is a good choice here and protects against null objects. The code as it current standards will accept Array's to be passed in.